plugins/inventory/foreman.py

```diff
@@ -238,10 +296,203 @@ def _get_hostvars(self, host, vars_prefix='', omitted_vars=()):
+        self.want_host_group = report_options['want_host_group']
+        self.want_hostcollections = report_options['want_hostcollections']
+        self.want_subnet = report_options['want_subnet']
+        self.want_subnet_v6 = report_options['want_subnet_v6']
+        self.want_smart_proxies = report_options['want_smart_proxies']
+        self.want_content_facet_attributes = report_options['want_content_facet_attributes']
+        self.want_host_params = report_options['want_host_params']
+        self.want_facts = self.get_option('want_facts')
+        self.host_filters = self.get_option('host_filters')
+
+        params["Organization"] = options[self.want_organization]
+        params["Location"] = options[self.want_location]
+        params["IPv4"] = options[self.want_IPv4]
+        params["IPv6"] = options[self.want_IPv6]
+        params["Facts"] = options[self.want_facts]
+        params["Host Group"] = options[self.want_host_group]
+        params["Host Collections"] = options[self.want_hostcollections]
+        params["Subnet"] = options[self.want_subnet]
+        params["Subnet v6"] = options[self.want_subnet_v6]
+        params["Smart Proxies"] = options[self.want_smart_proxies]
+        params["Content Attributes"] = options[self.want_content_facet_attributes]
+        params["Host Parameters"] = options[self.want_host_params]
+        if self.host_filters:
+            params["Hosts"] = self.host_filters
+        return params
+
+    def _use_inventory_report(self):
+        try:
+            use_inventory_report = self.get_option('foreman').get('use_reports_api')
+        except Exception:
+            use_inventory_report = False
+        if not use_inventory_report:
+            return False
+        status_url = "%s/api/v2/status" % self.foreman_url
+        result = self._get_json(status_url)
+        foreman_version = (LooseVersion(result.get('version')) >= LooseVersion(self.MINIMUM_FOREMAN_VERSION_FOR_REPORTING_API))
+        return foreman_version
+
+    def _post_request(self):
+        url = "%s/ansible/api/v2/ansible_inventories/schedule" % self.foreman_url
+        session = self._get_session()
+        params = {'input_values': self._fetch_params()}
+        try:
+            self.poll_interval = self.get_option('report').get('poll_interval')
+        except Exception:
+            self.poll_interval = 10
+        ret = session.post(url, json=params)
+        if not ret:
+            raise Exception("Error scheduling inventory report on foreman. Please check foreman logs!")
+        url = "{0}/{1}".format(self.foreman_url, ret.json().get('data_url'))
+        response = session.get(url)
+        while response:
+            if response.status_code != 204:
+                break
+            sleep(self.poll_interval)
+            response = session.get(url)
+        if not response:
+            raise Exception("Error receiving inventory report from foreman. Please check foreman logs!")
```

Comment:
this loop needs some kind of timeout, as otherwise if the api does not work for some reason (= does not return anything but `204`), we'll end up in and endless loop.

Reply:
We discussed this briefly when migrating to reports with tower initially and the time to generate the report is dictated by the number of hosts to be captured and both the report generation in foreman and inventory sync in tower are async jobs. Given that, should we add a (user-specified?) timeout in the script itself?

Reply:
While this is true for Tower, it's not for Ansible itself, here the inventory fetch is synchronous.

The case where I hit the issue was: my dynflow workers died, so while the request to generate an inventory was accepted, it was never processed, the api kept returning 204 and nothing worked.

Is there a guesstimate for large deployments? Ist 5 minutes a good timeout?

Partially unrelated: this triggers a regeneration on each call - is there a way to use an older report?

Reply:
Vladimir had done some load testing of this on the tower side initially. His tests showed roughly 3 minutes for 10k hosts. I think 10 minutes should be a good conservative timeout for waiting on the report generation to finish if we have that as default timeout..Pushed a change to add that.

I am not sure about the second question. Presently, If the report API fails, we fall back to the old host API.

Reply:
I should not write review comments on mobile… Sorry :)

> Partially unrelated: this triggers a regeneration on each call - is there a way to use an older report?

What I meant is: is there a way (in the API) to say "give me the last rendered version of this report, without generating a new one"? I can imagine on bigger environments, people would like to just render it on a timer, and then use the pre-rendered version instead of triggering a fresh render every time ansible runs. But this is totally out of scope here, so let's get on with the real thing.